File: LeetCode739/LeetCode739.java
import java.util.*;

class Solution {
    public int[] dailyTemperatures(int[] T) {

        Stack<List<Integer>> container = new Stack<List<Integer>>(); 
        int [] result = new int[T.length];
        container.push(Arrays.asList(0, T[T.length-1]));
        int counter;

        for (int i = T.length - 2; i >= 0; i--)
        {
            counter = 1;

            while (!container.empty() && T[i] >= container.peek().get(1))
            {
                counter += container.peek().get(0);
                container.pop();
            }

            if (container.empty())
            {
                container.push(Arrays.asList(0, T[i]));
            }
            else
            {
                container.push(Arrays.asList(counter, T[i]));
                result[i] = counter;
            }
        }

        return result;
    }
}

File: LeetCode739/LeetCode739.cpp
#include <vector>
#include <algorithm>

class Solution
{
public:
    std::vector<int> dailyTemperatures(std::vector<int> &T)
    {
        std::vector<int> result(T.size(), 0);
        std::vector<std::pair<int, int>> container;

        for (int i = 0; i < T.size(); i++)
        {
            while (!container.empty())
            {
                if (T[i] > container.back().second)
                {
                    result[container.back().first] = i - container.back.first();
                    container.erase(container.end() - 1);
                }
                else
                {
                    break;
                }
            }
            container.push_back(std::make_pair(i, T[i]));
        }

        return result;
    }
};